intrahost.py

```diff
@@ -129,157 +169,211 @@ def parser_tabfile_rename(parser=argparse.ArgumentParser()):
+                        s_chrom = samp_to_cmap[s].mapBtoA(ref_sequence.id)
+                        if row[0] == s_chrom:
+                            row = {'sample':s, 'CHROM':ref_sequence.id,
+                                's_chrom':s_chrom, 's_pos':int(row[1]),
+                                's_alt':row[2], 's_ref': row[3],
+                                'n_libs':row[7], 'lib_bias': row[8],
+                                'alleles':list(x.split(':') for x in row[9:] if x)}
+                            # make a sorted allele list
+                            row['allele_counts'] = list(sorted(
+                                [(a,int(f)+int(r)) for a,f,r in row['alleles']],
+                                key=(lambda x:x[1]), reverse=True))
+                            # reposition vphaser deletions minus one to be consistent with
+                            # VCF conventions
+                            if row['s_alt'].startswith('D'):
+                                for a,n in row['allele_counts']:
+                                    assert a[0] in ('D','i')
+                                row['s_pos'] = row['s_pos']-1
+                            # map position back to reference coordinates
+                            row['POS'] = samp_to_cmap[s].mapAtoB(s_chrom, row['s_pos'], side = -1)[1]
+                            row['END'] = samp_to_cmap[s].mapAtoB(s_chrom, row['s_pos'], side = 1)[1]
+                            data.append(row)
+            
+                # sort all iSNVs (across all samples) and group by position
+                data = sorted(data, key=(lambda row: row['POS']))
+                data = itertools.groupby(data, lambda row: row['POS'])
+            
+                # process one reference position at a time
+                for pos, rows in data:
+                    rows = list(rows)
+                
+                    # define the length of this variation based on the largest deletion
+                    end = pos
+                    for row in rows:
+                        end = max(end, row['END'])
+                        for a,n in row['allele_counts']:
+                            if a.startswith('D'):
+                                # end of deletion in sample's coord space
+                                local_end = row['s_pos']+int(a[1:])
+                                # end of deletion in reference coord space
+                                ref_end = samp_to_cmap[row['sample']].mapAtoB(row['s_chrom'], local_end, side = 1)[1]
+                                end = max(end, ref_end)
+                
+                    # find reference allele and consensus alleles
+                    refAllele = str(ref_sequence[pos-1:end].seq)
+                    consAlleles = {} # the full pos-to-end consensus assembly sequence for each sample
+                    samp_offsets = dict((row['sample'], row['s_pos']) for row in rows) # the isnv's index in the consAllele string for each sample
```

Comment:
I *think* I'm handling this scenario but we should add an explicit unit test for it... I don't think the result should depend on the order of the input...

--
Daniel J. Park, PhD
Group Leader, Viral Computational Genomics
Broad Institute of MIT and Harvard
Tel: +1-617-714-8526
dpark@broadinstitute.org
http://www.broadinstitute.org/bios/daniel-park

> On Mar 2, 2015, at 5:52 PM, iljungr <notifications@github.com> wrote:
> 
> In intrahost.py:
> 
> > +                    # define the length of this variation based on the largest deletion
> > +                    end = pos
> > +                    for row in rows:
> > +                        end = max(end, row['END'])
> > +                        for a,n in row['allele_counts']:
> > +                            if a.startswith('D'):
> > +                                # end of deletion in sample's coord space
> > +                                local_end = row['s_pos']+int(a[1:])
> > +                                # end of deletion in reference coord space
> > +                                ref_end = samp_to_cmap[row['sample']].mapAtoB(row['s_chrom'], local_end, side = 1)[1]
> > +                                end = max(end, ref_end)
> > +                
> > +                    # find reference allele and consensus alleles
> > +                    refAllele = str(ref_sequence[pos-1:end].seq)
> > +                    consAlleles = {} # the full pos-to-end consensus assembly sequence for each sample
> > +                    samp_offsets = dict((row['sample'], row['s_pos']) for row in rows) # the isnv's index in the consAllele string for each sample
> If consensus for a sample has an insertion relative to reference, then can't several values of s_pos in that sample map to the same POS? For example:
> Ref: A--A
> Cons: AAAA
> Allele1: ATAA, POS=1, s_pos=2
> Allele2: AATA, POS=1, s_pos=3
> In that case, when we form this dictionary we would be adding several s_pos values with the same sample value and the result will depend on the order in which they are added.
> 
> —
> Reply to this email directly or view it on GitHub.
>